reduce diagnostic tagger's usages of UI thread for synchronization

heejaechang opened the pull request:
**Customer scenario**

Customer is working on a VS. doing code fix and watching previews, applying fix all and etc, and suddenly VS crash due to OOM.

**Bugs this fixes:**

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/512757

**Workarounds, if any**

there is no easy workaround.

**Risk**

I don't see risk of crash, but since it is highly visible area (tagger), we probably want more dogfooding time for this.

**Performance impact**

it should move more work to background thread, and reduce need for UI thread. in turn, improving responsiveness and memory.

**Is this a regression from a previous update?**

Yes.

**Root cause analysis:**

this one is hard to pin point to 1 issue. but caused by several different fixes. 

first, we made all state management to happen on UI thread to remove any potential race. and that caused us to use more FG than before.

second, we merged preview only tagger with regular diagnostic tagger, making preview tagger to be as expansive as normal diagnostic tagger meaning it uses same amount of UI threads since it now has same states tracking as normal ones.

third, along with that, some filtering on BG got removed which was there to reduce work on FG since all those are changed to be done on FG. 

forth, diagnostic tagger didn't pass cancellation token all the way through, leaving already cancelled works in the UI work queue.

fifth, foreground work queue cleaned up cancelled works too lazily and let cancelled work to pile up.

..

this reduces demand for UI thread, but due to the fact that UI thread can be blocked (due to exclusive operation running such as Fix All, or model dialog box or Wait dialog box and more), we can't completely get rid of pending requests for UI threads. but with this change, we should have way less chance to hit 2.1 million pending UI requests.

**How was the bug found?**

MemoryWatson

heejaechang commented on src/Workspaces/Core/Portable/Workspace/Workspace_Registration.cs:
```diff
@@ -58,12 +55,27 @@ protected void UnregisterText(SourceTextContainer textContainer)
+            // guard us from being called with wrong text container.
             if (registration.Workspace == this)
             {
-                registration.SetWorkspaceAndRaiseEvents(null);
```
register text is async event, and unregister as synchronous events which can cause ordering issue (unregister fire first and then register event get fired) so moved all events to async

heejaechang commented:
retest this please

heejaechang commented:
@dotnet/roslyn-analysis @dotnet/roslyn-ide can you take a look?

CyrusNajmabadi commented:
It's very unclear to me what hte actual issue is here that is being addressed.  How did we get  2.1 million pending UI requests. in the first place, given how things are batched?

CyrusNajmabadi commented:
> first, we made all state management to happen on UI thread to remove any potential race. and that caused us to use more FG than before.

Yes.  How are we ensuring that we are not introducing race conditions here like we had before.  The move to the FG thread as an appropriate sync context was *very intentionally* done because of how many race conditions existed prior to that change.

> second, we merged preview only tagger with regular diagnostic tagger, making preview tagger to be as expansive as normal diagnostic tagger meaning it uses same amount of UI threads since it now has same states tracking as normal ones.

I don't know what this means.  "it uses hte same amount of UI threads".  There should only be one UI thread.  Why would the preview tagger have a particular problem here?

> third, along with that, some filtering on BG got removed which was there to reduce work on FG since all those are changed to be done on FG.

Again, this was due to race conditions.

> forth, diagnostic tagger didn't pass cancellation token all the way through, leaving already cancelled works in the UI work queue.

Again, this was due to race conditions.  It became very intentional in the code at which points things could truly be canceled, vs getting to a point of no return wherin state changes had happened, and thus needed to fully be persisted in the appropriate places.

> fifth, foreground work queue cleaned up cancelled works too lazily and let cancelled work to pile up.

This seems like an issue that sounds safe to fix and could be fixed independently.

heejaechang commented:
@CyrusNajmabadi you can take a look how much FG requests diagnostics taggers puts in in simple Console app.

easily reach 10000 requests just cycling through preview window.

heejaechang commented:
@CyrusNajmabadi also, you didn't specify exactly what the race was, since it didn't specify, I have no idea whether there was other way to remove race instead of doing everything in UI thread.

heejaechang commented:
"I don't know what this means. "it uses hte same amount of UI threads". There should only be one UI thread. Why would the preview tagger have a particular problem here?"

I meant amount of UI thread requests. before, simple preview workspace tagger didn't need any of this complex state management, since each analyzer will return diagnostic only once and never change, so there was no race at all and no reason to use UI thread.

heejaechang commented:
@CyrusNajmabadi  "Again, this was due to race conditions." so, exactly what race?

heejaechang commented:
"Again, this was due to race conditions. It became very intentional in the code at which points things could truly be canceled, vs getting to a point of no return wherin state changes had happened, and thus needed to fully be persisted in the appropriate places."

well, view is going away and tagger is disposed. no idea when tagger goes away what race tagger can have.

heejaechang commented:
"This seems like an issue that sounds safe to fix and could be fixed independently."

could be, but right now diagnostic tagger is the only one who doesn't cancel work when work is already invalidated. so this change doesn't do anything without tagger fix since, all work from diagnostic tagger doesn't pass right cancellation token.

CyrusNajmabadi commented:
> you can take a look how much FG requests diagnostics taggers puts in in simple Console app.
> easily reach 10000 requests just cycling through preview window and close LB.

How is that happening?  what are those 10k requests?  If it's just a console app, how do we have that much going on?  I think we need to figure out what's going on in the system instead of trying to fix a symptom here.

Even if we make things do more on the BG, we're still doing 10k things on the BG, despite there not seeming to be a good reason for that at all.

CyrusNajmabadi commented:
> also, you didn't specify exactly what the race was, since it didn't specify

The race conditions were the issues that caused us to have things like squiggles that never went away.  Because we were not vigilant about keeping our data store safe, and we were haphazardly accessing it and mutating it from several threads, we would end up with incorrect data, and improper presentation of that data.

We moved to a lockstep approach for the diagnostic tagging system where we would carefully serialize out mutations and we would then ensure that once mutations occurred we would always ensure that the tagging system was then updated to reflect the delta changes that we had recorded.

Note: as mentioned in another bug today, i think it's *very unfortunate* that any of this code exists in the tagging layer at all.  The core problem is that the tagging system has had to backward infer the contract about how diagnostics work in an attempt to actually understand and report what the state of the system is at any point in time for any buffer.  As we've seen in the past there have been numerous events fired that it didn't know it was supposed to listen to.  And there have been many complexities about diagnostics that it was unaware of that caused it to have a different belief about what diagnostics were around, versus what the actual diagnostics subsystem thought there should be.

--

IMO, due to how buggy this entire space has been (for 5+ years now), it really should be rewritten with a far more appropriate architecture.  Specifically, instead  of having hte tagger try to talk to low level storage and eventing systems, it should talk to a higher level service that it registers and unregisters documents of interest with.  This service, exposed at the diagnostics subsystem layer, would properly aggregate diagnostics for that document, and push normal events when they changed.  The diagnostic tagger would then be a super simple tagger (like the rest of our taggers).  

This would isolate the complexity, instead of having it woven into the tagging infrastructure.  If you wanted to make it all done in the BG for that new service, you could totally do that.  Then, all this mess of BG/FG interactions in the tagger would not be a problem.

CyrusNajmabadi commented:
> could be, but right now diagnostic tagger is the only one (2.1 million requests, almost all of them is from diagnostic tagger)

Why are there 2.1 million requests from the diagnostic tagger?  What is causing that to happen?

CyrusNajmabadi commented:
> all work from diagnostic tagger doesn't pass right cancellation token.

Which cancellation tokens are you talking about, and to which systems is it not passing the token correctly.  There are very specific cancellation tokens in the tagger for very specific purposes.  Including different cancellation tokens to deal with needing to get the first set of diagnostic tags and whatnot.  There are ordering concerns between these different parts, and getting it wrong can produce races.  For example, it's possible, if your'e not careful, to process a "diagnostics-are-changing" event before processing the "these-are-the-initial-diagnostics" event.  So we have to be careful here.

--

My ideal would be to separate these out entirely.  For one, it would help actually simplify the code majorly.  For another, it would mean that each system has clear responsibility.  i.e. the diagnostics tagger code would not have to think about the threading concerns.  And the diangostics aggregation code would not have to thnk about tagging concerns.  

Right now, throwing more complexity into an already complex system is not a viable path forward.  Even if we're lucky enough to get it correct (which i'm not convinced of based on my reading of the PR), it will be a very costly system to maintain moving forward.  Most likely outcome it just having another set of racey diagnostics bugs appear, and having to revisit this problem 6 months down the line.  

--

The rest of the taggers are effectively a solved space.  They've all be working reliably and properly for 5ish years+ at this point.  It would be nice to get diagnostic tagging into the same space.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -56,13 +57,22 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
                 // Kick off a background task to collect the initial set of diagnostics.
-                var cancellationToken = _initialDiagnosticsCancellationSource.Token;
                 var asyncToken = _owner._listener.BeginAsyncOperation(GetType() + ".GetInitialDiagnostics");
-                var task = Task.Run(() => GetInitialDiagnosticsInBackground(document, cancellationToken), cancellationToken);
```
not passing this token is not good.

heejaechang replied:
it is field. doesn't need to pass in. it can use field to check.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -74,21 +84,21 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
-                        var diagnostics = AdjustInitialDiagnostics(project.Solution, updateArgs, cancellationToken);
-                        if (diagnostics.Length == 0)
+                        var diagnostics = AdjustInitialDiagnostics(project.Solution, updateArgs, _shutdownCancellationSource.Token);
+                        if (diagnostics.Length == 0 || !diagnostics.Any(d => _owner.IncludeDiagnostic(d)))
```
what is this code for?

heejaechang replied:
so that unnecessary tagger doesn't enqueue task for diagnostics that are not related to unnecessary tagger.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -190,19 +200,25 @@ private void DisconnectFromAllTaggers()
                 {
+                    if (e.DocumentId != _currentDocumentId)
+                    {
+                        // ignore diagnostics that belong to different documents
```
is this racey?  what happens if our buffer now maps to a different document?  will we suddenly ignore events about the diagnostics for the buffer before the document changed?

heejaechang replied:
no. it won't. if it does, then that is an issue in Workspace.Events it is listening to.

heejaechang replied:
if there is an issue in Workspace events, and if we are trying to fix that by moving everything on UI thread and do filtering there, that is exactly trying to fixing the symptom not root cause of it.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -213,6 +229,36 @@ private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
+            {
+                this.AssertIsForeground();
+
+                // reset currentDocumentId if needed
```
comment repeat what code says.

heejaechang replied:
sure.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -236,18 +282,12 @@ private void OnDiagnosticsUpdatedOnForeground(DiagnosticsUpdatedArgs e)
                     return;
                 }
 
+                _shutdownCancellationSource.Token.ThrowIfCancellationRequested();
```
you're on the UI thread.  if you throw an exception, what happens?  what is higher in the stack to catch this?

heejaechang replied:
this runs in foreground notification service, it will be caught and ignored.

jasonmalinowski replied:
(I think I've already caught bugs masked by that foreground notification service behavior. Would prefer exceptions just took down process, cancellation included, if exceptions leak that far.)

heejaechang replied:
so, should I throw or just return if cancelled?

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -364,6 +418,8 @@ private void RemoveAllCachedDiagnostics()
                 this.AssertIsForeground();
                 Debug.Assert(!_disposed);
 
+                _shutdownCancellationSource.Token.ThrowIfCancellationRequested();
```
same concern.  should our UI thread checks actually just be returning instead of throwing.

heejaechang replied:
same here. sure, we can change it to return.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Workspace/Workspace_Registration.cs:
```diff
@@ -58,12 +55,27 @@ protected void UnregisterText(SourceTextContainer textContainer)
+            var oldWorkspace = registration.Workspace;
+
+            // set workspace right away, but events are lazy. 
+            // people, who is checking Registration.Workspace directly, will see the change right away but
```
who does this though?   Is there a benefit to not just making the entire thing lazy?

heejaechang replied:
I am not sure I am following you. this is how whole Roslyn Workspace eventing works.

sharwell commented:
:memo: For users impacted by the underlying bug, the maximum observed case is 21 million items in the queue, and the *average* number of items is 2.4 million.

sharwell commented:
> It's very unclear to me what hte actual issue is here that is being addressed. How did we get 2.1 million pending UI requests. in the first place, given how things are batched?

This is the hardest part here. MemoryWatson is very specific about where we ended up, but doesn't provide very much information at all about how we got there.

jinujoseph commented:
@heejaechang , can you pls retarget this for post 15.5 branch pls.

heejaechang commented:
the root cause of the complexity comes from the fact that 

1. tagger is associated to a buffer. not document. 
2. and diagnostics associated to document not on buffer.
3. buffer can attached and detach to different workspace and documents.

so, diagnostic service doesn't care about buffer and raise events on (analyzer, document, diagnostics). sure, we can add one more layer that raise events on (analyzer, diagnostics) by filtering on documents given, but that doesn't remove the root cause of the complexity. that filtering can happen in any layer.

but regardless, diagnostic service should never care about buffer since at the layer it is working on, document is the unit it cares.

since this particular tagger is the one who wants to get diagnostics of the document this buffer is at the point associated with, the tagger should be the one who manage (3).

existing code basically gave up on managing (3) and choose most expansive way to do it. get all events to foreground threads, and it does the all filtering at the UI thread. 

this change tries to move that down to BG as much as possible. so we don't need to enqueue work to UI for nothing.

jasonmalinowski commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -56,13 +57,22 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
                 _owner = owner;
                 _subjectBuffer = subjectBuffer;
 
+                // track context changes
```
We were never doing this before? Was this scenario always broken?

heejaechang replied:
we did this by checking last documentId with what document buffer is associated with at foreground thread when we actually update state. rather than when context changed reported by workspace.

CyrusNajmabadi replied:
This was important due to racey interleavings of events from the diagnostics service and the workspace service.  I'm not convinced this won't just lead to a problem where tagger diagnostics beocme out of sync.

jasonmalinowski commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -56,13 +57,22 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
 
+                // track context changes
+                _registration = Workspace.GetWorkspaceRegistration(_subjectBuffer.AsTextContainer());
+                _registration.WorkspaceChanged += OnWorkspaceChanged;
```
This isn't unsubscribed from and thus a leak.

heejaechang replied:
when tagger goes away, registration also goes away, so, it won't be leak, but sure, I can be explicit and unsubscribe at the dispose.

jasonmalinowski commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -56,13 +57,22 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
-                var cancellationToken = _initialDiagnosticsCancellationSource.Token;
                 var asyncToken = _owner._listener.BeginAsyncOperation(GetType() + ".GetInitialDiagnostics");
-                var task = Task.Run(() => GetInitialDiagnosticsInBackground(document, cancellationToken), cancellationToken);
+                var task = Task.Run(() => GetInitialDiagnosticsInBackground(document), _shutdownCancellationSource.Token);
```
So what happens if the document ID changes while this is running? I see us cancelling it, but I don't see how we'd start the new request for the correct document?

heejaechang replied:
turns out this made it work. 
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs,58

we lost history so, not sure why that is added. but have feeling it probably related to closed file analysis option. I believe now all documents of same buffer are open or closed together, so I think this won't be needed any more. 

anyway, I am going to remove the above since diagnostic shouldn't care about the context change, and make tagger to deal with context change.

jasonmalinowski commented on src/EditorFeatures/Core/Implementation/ForegroundNotification/ForegroundNotificationService.cs:
```diff
@@ -311,9 +323,34 @@ private void Enqueue_NoLock(LinkedListNode<PendingWork> entry)
-                    current = current.Previous;
+                    var previous = current.Previous;
+
+                    // eagarly clean up cancelled entry from the list backward
```
Why are we cleaning up the queue during a later enqueue versus cleaning it up eagerly when the cancellation was triggered?

heejaechang replied:
are you saying we should register action for each cancellation? that's one way.. but that probably require some thought on concurrent issue since cancellation can happen in any thread. 

here I am cleaning up very naively from BG. work that got cancelled might still left in the queue and cleaned up when running in FG. 

removing 10K+ items from list from FG in 15ms is not an issue if that ever happens.

CyrusNajmabadi replied:
> removing 10K+ items from list from FG in 15ms is not an issue if that ever happens.

This is fixing a symptom.  Why are we getting 10k items in the FG queue.  You mentioned it happens just with a simple console app.  Certainly it should be possible to investigate why that is happening and fixing the root cause.

heejaechang replied:
rest of PR is fixing the issue we are getting thousands of UI requests. and I already stated that was happening since tagger was moving everything to UI thread without filtering anything on BG.

there was even more I could improve here since tagger can be disposed while its request is on UI thread which can not be discarded. and this make general cancellation cleanup better for this service.

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs:
```diff
@@ -41,10 +42,16 @@ private DiagnosticIncrementalAnalyzer GetOrCreateIncrementalAnalyzer(Workspace w
+                // workspace such as previewWorkspace, which can come and go on the fly, will use this to remove data saved. 
+                // but since this doesn't raise diagnostic removed events, if there are others
+                // who hold onto diagnostics reported, it is their responsibility to clean them up when workspace
+                // goes away
```
I don't understand this comment.

heejaechang replied:
since this raise events, any one can hold onto diagnostics reported. which has pointer to Workspace.

before, it made sure it raises events saying, this diagnostics are going away. so who listening the diagnostic service, can clean them up. now, I changed it to not raise. so if there is any (currently there is no one), they should care about Workspace, and if Workspace goes away, they should drop diagnostics they are holding to, otherwise, it will leak Workspace.

for this particular tagger, since it goes away, when Workspace goes away, not dropping diagnostics it is holding onto is fine.

CyrusNajmabadi replied:
None of this is clear in the code.  there are complex lifetime and eventing scenarios going on here, and from otehr changes you've made, it seems like you're changing things inside the diagnostics service itself.  For example, you said "anyway, I am going to remove the above since diagnostic shouldn't care about the context change".

This means the tagger code needs to have intimate knowledge about how the diagnostics service behaves, and when the diagnostics service changes, it can have deeply subtle implications for code like this.  

That makes the coupling here very brittle and difficult to maintain.  Effectively, the tagging code needs to be very familiar with how the diagnostics service is implemented, and any chnages to the latter need to be carefully absorbed into the former.

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/DiagnosticService_UpdateSourceRegistrationService.cs:
```diff
@@ -31,5 +32,23 @@ public void Register(IDiagnosticUpdateSource source)
+        public void Shutdown(IDiagnosticUpdateSource source, Workspace workspace)
+        {
+            var eventToken = _listener.BeginAsyncOperation(nameof(Shutdown));
+            _eventQueue.ScheduleTask(() =>
```
Can you add a comment explaining why this is being done in a deferred way?

heejaechang replied:
sure.

heejaechang commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -56,13 +57,22 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
+
+                if (_registration.Workspace != null)
+                {
+                    // in unit test, it is possible, we create tagger for buffer which doesn't have associated document
```
yep. turns out we do. I didn't pay much attention why not, but I can remove this and see why those unit tests are not setting registration.

jasonmalinowski replied:
Please do, this seems insane.

jasonmalinowski replied:
Now, mind you, it's entirely possible for our tagger to exist on a buffer with no workspace -- that's a scenario that's not unit test at all. So maybe just remove the comment?

heejaechang commented on src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticTaggerWrapper.cs:
```diff
@@ -135,6 +136,15 @@ public void Dispose()
                 _solutionCrawlerService.WaitUntilCompletion_ForTestingPurposesOnly(_workspace, _incrementalAnalyzers);
             }
 
+            // if workspace waiter exist, wait for it first.
```
sure

heejaechang commented on src/Workspaces/Core/Portable/Workspace/WorkspaceRegistration.cs:
```diff
@@ -15,20 +15,26 @@ internal WorkspaceRegistration()
         }
+    }
+
+    internal sealed class WorkspaceChangedEventArgs : EventArgs
```
sure

heejaechang commented on src/Workspaces/Core/Portable/Workspace/Workspace_Registration.cs:
```diff
@@ -58,12 +55,27 @@ protected void UnregisterText(SourceTextContainer textContainer)
             }
         }
 
+        private void RaiseTextRegistrationChangedEvents(WorkspaceRegistration registration, Workspace newWorkspace, [CallerMemberName] string eventName = null)
```
sure.

CyrusNajmabadi commented:
> so, diagnostic service doesn't care about buffer and raise events on (analyzer, document, diagnostics). sure, we can add one more layer that raise events on (analyzer, diagnostics) by filtering on documents given, but that doesn't remove the root cause of the complexity. that filtering can happen in any layer.

I firmly disagree with this.  This would absolutely remove complexity by encapsulation and modularization.  It would enable us to expose a very clean API with appropriate understanding at appropriate layers.

It would also enable testing of just that part of the system.  Finally, it would give a much cleaner and more understandable threading model.  If you wanted, say, to avoid using the UI thread for synchronization (which i know is something you've complained about in the past), you could do things entirely with bg sync, but in a module owned a the diagnostics layer with appropriate understanding of the threading rules the diagnostics subsystem and workspace layer exposed.

The tagger infrastructure could continue using the existing *effective* threading model that it has today which provides simplicity and very strong consistency.  Something very important for a system which needs to be long running, interact with the UI layers of VS, and must always reach a correct fixed point against the users' code.

CyrusNajmabadi commented:
> but regardless, diagnostic service should never care about buffer since at the layer it is working on, document is the unit it cares.

I have not asked for the diagnostic service to care about buffers.  I've asked for there to be a component provided that properly understands the connection between the to and encapsulates that understanding in a single place that can then be properly documented, tested, and maintained.  Right now there is *zero* documentation *anywhere* about how a system (like tagging) should use the diagnostic service, and what sort of *other* components it must also integrate with and listen to in order to use the diagnostics service properly.  For example, over the years there have been many events that that exist that cause the diagnostics service to do things, but which don't end up causing events to bubble up through the event it itself fires.  So consumers of the diagnostics service need to know aobut those events and then know what that means for any existing data they consumed from the diagnostics service, as well as any future data that they then may encounter.  And, because none of this is documented (including any of the threading concerns around all these different streams of events) it is very difficult for such layers to know if there are implicit race conditions they need to know about.

*For example*

Say something changes in the workspace that would cause a buffer to be associated with a different document.  It is possible for this layer to hear about events in all sort of different orders including (note: my memory may be off here, but these are the sorts of things that can happen): 

1. Hearing from teh workspace first that things have changed.
2. Then hearing about changes from the diagnostics layer about the new document.
3. Then hearing about there being no diagnostics from the old document.

However, that order is not guaranteed,  It is possible to get many different reorderings of the above.  So, for example, it might happen that the system hears about the removal of the old doucment, and the introduction of hte new document, but it hasn't heard from teh workspace yet that the doc changed.  As such, it can "lose" the knowledge about the initial set of diagnostics for the new document.

Currently, the only way for the tagging systme to try to get this all right, is by deeply introspecting the *implementation* of the diagnostics service.  Then, once that has been observed, it is necessary to try to log and then piece back all the potential event orderings (hoping that none were missed).  

This is extremely error prone and complex.  And that only gets worse when complex new threading models, asynchrony, and cancellation are added into the mix.

heejaechang commented:
@CyrusNajmabadi IDiagnosticService should not know about buffer. tagger is the one associated with buffer not diagnostic service. Roslyn has 2 different documents for the buffer. it is buffer that is shared not document.

buffer's workspace changes or buffer context changes, doesn't require any change in diagnostics.

heejaechang commented:
@CyrusNajmabadi current tagger is subscribed to events host wide events and move all that events to UI thread and filter out all events that are not needed. that is why UI threads got hundreds of events.

I added another events that does filtering inside. changed tagger to use new event. same code, just where code exist moved.

CyrusNajmabadi commented:
>that is why UI threads got hundreds of events.

Hundreds of events is fine.  How are we getting 21million objects out of that that have to be cleaned up?

CyrusNajmabadi commented:
> I added another events that does filtering inside.

Yes.  I get that.  My problem is that i'm not convinced this is race free.  You're filtering using data that may not be valid on that thread based on what's happening with other events.  For example the workspace events that talk about how buffers/documents are chaning.  That's why we have synchronization points.  So that the data is always examined in a serial fashion from a context that can then ensure that it's always consistent.

heejaechang commented:
> Hundreds of events is fine. How are we getting 21million objects out of that that have to be cleaned up?

I meant console with 1 files. roslyn that has 10000 files, with 100 also analyzers can create million events if every single document has errors for every single analyzers. and existing code was moving all those events blindly to UI thread and then throw away diagnostics from 9999 files.

CyrusNajmabadi commented:
> IDiagnosticService should not know about buffer.

I have never asked for IDiagnosticServce to know about buffer.

I have consistently and repeatedly asked for a component to be provided that encapsulates the knowledge of the data provided from IDiagnosticService along with knowledge of all the events that can affect it wrt the workspace, and packages that up in an API that consumers can effectively use so that they don't need to know the internal implementation details of IDiagnosticService.   Right now there is a deeply brittle dependency between the two.  With a lot of conservative code needed to try to provide correctness in the face of such undocumented behavior.   

Right now Diagnostics are literally the only place in our system that has this issue.  All other tagger systems avoid it because they sit on subsystems that abstract this all away and provide a sensible facility for the tagging system to know when it should rerun, and how it should acquire data when it needs to rerun.  These patterns are used extensively across roslyn precisely because it allows clean encapsulation (which also has benefits for testing), and decoupling of concerns.  I'm asking that we move to something similar to that for diagnostics so that squiggles does not need to be in the business of knowing exactly how the diagnostics service was implemented, and all the complex cases is needs to handle there.

--

Not to beat a dead horse on this.  But this coupling has been a problem for years at this point.  Unlike most of our other taggers (which have effectively been solved and stable for years), squiggles continues to face problems due to thsi problem.  It's my opinion that if we take this sort of fix that we'll just end up having to revisit this 6 months down the line.  At this point, to me, it's really time to just do this properly and implement a proper solution that solves this once and for all instead of just continually trying to fix things by adding and shuffling around complexity.

heejaechang commented:
@CyrusNajmabadi we have 2 asynchronous events and 1 synchronous events. one from workspace, one from diagnostic service, and one from WorkspaceRegistration.

tagger needs to stitch them together. it has nothing to do with how diagnostic service works, it is basic problem of when you have multiple event sources mixing async and synchronous events, how you are going to stitch them together.

moving all those to UI thread doesn't fix anything, since it doesn't solve the problem of how to order those events. one doesn't know whether workspace events it got (context changed) happened before WorkspaceRegistration.WorkspaceChanged events, or DiagnosticUpdate events happened before or after the other events.

many things needed to be used together to mix those events properly.

right now, it will use taskChain to serialize all events to one timeline. record context change (workspace, contenxt change) on UI thread and when it happens, refresh everything from scratch.

when it refresh from scratch, it will first subscribe to diagnostic changed events, and pull current data stored in the service. 

sure it can be out of sync, right after context changes until all async events are conumsed, but at the end, it should be in sync with Workspace.CurrentSolution.

CyrusNajmabadi commented:
> and existing code was moving all those events blindly to UI thread and then throw away diagnostics from 9999 files.

Again, it was not blindly.  It was *specifically* to avoid race problems where events were missed because the different processing was happening on different threads, and filtering like this was either missing events, or was processing the wrong events because it was running concurrently with workspace changes.

--

At this point i don't know what more i can say on this point.  I firmly understand the problem you're trying to solve.  And i see an effective means of addressing said problem.  But i feel like instead of taking that route (proper encapsulation around a documented and testable API) this PR is taking an extremely risky approach of just shifting things around in a complex fashion, opening us to race conditions under the guise of improved performance.  Everythign i know about this code simply tells me that if we take this approach, we'll have to make another complex change here months down the line.  

I would *strongly* advise just going and writing that other component.  Honestly, i don't even think it would be that hard.  I would even be willing to try to make time to write it myself this weekend, but i can't guarantee i'll have the time.  Critically, it would help reduce complexity and provide a centralized location for people to review the most critical and complex parts around threading and asynchrony.  

These sorts of approaches are virtually always what we want for these sorts of system's coupling problems, and it's what we've effectively done elsewhere to great effect.  It allows for much easier understanding of what the core data is, how it is locked, how it mutates, and how it aggregates and notifies relevant consumers.  Furthermore, it allows the shared knowledge to be kept together.  i.e. if we change things in IDiagnosticSerivce, we can limit changes to here, instead of having it leak out into unrelated systems like tagging.

CyrusNajmabadi commented:
> I meant console with 1 files.

Previously you said this:

> you can take a look how much FG requests diagnostics taggers puts in in simple Console app.
> easily reach 10000 requests just cycling through preview window and close LB.

How does that happen?  What is causing a console app to generate 10k requests just when you open the preview window?

CyrusNajmabadi commented:
> it has nothing to do with how diagnostic service works, it is basic problem of when you have multiple event sources mixing async and synchronous events, how you are going to stitch them together.

your first clause contradicts the second.  You need to know how the diagnostic service sends its own events.  Something which is *not documented* anywhere.  When i did some of the work here previously, i was only able to figure it out by going through the code for the impl of it and seeing how it worked.

CyrusNajmabadi commented:
> but at the end, it should be in sync with Workspace.CurrentSolution.

I don't see how you are ensuring that with your PR.

heejaechang commented:
@CyrusNajmabadi because preview window basically has 2 views and it had about 8-10 also fixes in it. create property, type, nuget and etc.

so you have 2 * 8 = 16 preview workspaces, which all has 1 documents in it. it listens to analyzers, 10+ I think?

now, IDiagnosticService analyzers are host wide. meaning it fires events for any workspace for any document for any analyzers.

so, basically there are 160 events assuming there is only 1 diagnostics per analyzer. and there are 3 tagger that listens to this events. so now it is 420 events. that all goes to UI thread just so that, probably all of them discarded since it doesn't belong to right workspace, or not right diagnostics for this tagger. in this case there was only 1 doc, but if there was more, then it also will do filtering on doc on UI thread.

and I cycled through those preview windows meaning, same events fires again and again piled up.

CyrusNajmabadi commented:
> it doesn't solve the problem of how to order those events.

i agree. which is why i've been saying we need a solution that does precisely this.  This is, in fact, exactly what i've been asking for for several years :)

We have this with (i think) all our other subsystems that compute and cache data.  It's why, for example, none of our other taggers have had to deal with this, and why they haven't had these sorts of issues.  We have a lot of experience designing components that provide an appropriate API for accessing data in an concurrent world with asynchronous events happening.  Diagnostics are (afaict) the *only* system we have that doesn't follow the example that we have for the rest of Roslyn.  If we were to put in the effort to get it to do so, we could likely address the problem here, the reports about race conditions, and a whole swath of future bugs, in one fell swoop IMO.

heejaechang commented:
> your first clause contradicts the second. You need to know how the diagnostic service sends its own events. Something which is not documented anywhere.

I am not sure which one is contradicts the second. sure I could have better documentation, but can't understand why that is source of everything bad.

if that is the case, then WorkspaceRegistration.WorkspaceChanged would be more sever since it is public API, and no documentation and it actually could fire events in wrong order.

CyrusNajmabadi commented:
> so now it is 420 events

How did you go from 420 events to 10k events?

heejaechang commented:
> but at the end, it should be in sync with Workspace.CurrentSolution.
> I don't see how you are ensuring that with your PR.

so, how the existing code ensured it?

heejaechang commented:
"Diagnostics are (afaict) the only system we have that doesn't follow the example that we have for the rest of Roslyn"

I really don't understand what you mean by this. give me the example.

heejaechang commented:
> How did you go from 420 events to 10k events?

cycle through previews.

CyrusNajmabadi commented:
>  IDiagnosticService analyzers are host wide. meaning it fires events for any workspace for any document for any analyzers.

Maybe if we could have something not host-wide that would be good :)  

> if that is the case, then WorkspaceRegistration.WorkspaceChanged would be more sever since it is public API, and no documentation and it actually could fire events in wrong order.

Sure.  But having multiple bad things going on here doesn't mean we shouldn't be pushing for lots of improvements.  I don't think DiagnosticService is entirely at fault here.  But, i would also point out that we haven't had nearly as many issues (especially complex racey issue with enormous surrounding complexity) with WorkspaceChanged.  If we were having the same sorts of issues with it then i would *definitely* be recommending that we do something to improve it.

CyrusNajmabadi commented:
> I really don't understand what you mean by this. give me the example.

Er.. literally none of the other taggers or presentation systems we have need to do this sort of thing.  Squiggle tagging literally is "the example" :)

heejaechang commented:
> i would also point out that we haven't had nearly as many issues (especially complex racey issue with enormous surrounding complexity) with WorkspaceChanged

well, using WorkspaceChanged events without mixing other events sure. using Diagnosticservice events itself shouldn't give any trouble as well.

problem is when someone tries to mix WorkspaceChanged events with DiagnosticService events. 2 are both async events, there are no central coordinator that coordinate between 2 async events.

so, mixing these 2 are issue. I agree I just don't understand why diagnostic service's responsibility to do that? it is completely different thing or ones that want to mix 2 are in charge, in this case, the tagger.

CyrusNajmabadi commented:
> so, how the existing code ensured it?

For one thing, it *definitely* did not try to work across workspace registration changes.

CyrusNajmabadi commented:
>  I agree I just don't understand why diagnostic service's responsibility to do that?

I have never asked for this.  Please stop saying so.  Indeed, i just spent time above to reiterate that i'm not asking or this.

heejaechang commented:
> For one thing, it definitely did not try to work across workspace registration changes.

which is a bug? who ensure tagger to be refreshed? it subscribe to workspacechange event source for tagger, but it never refreshed tags. so it still shows old tags that belong to wrong workspace.

CyrusNajmabadi commented:
> problem is when someone tries to mix WorkspaceChanged events with DiagnosticService events. 2 are both async events, there are no central coordinator that coordinate between 2 async events.

Yes. I just said that above:

> I have consistently and repeatedly asked for a component to be provided that encapsulates the knowledge of the data provided from IDiagnosticService along with knowledge of all the events that can affect it wrt the workspace, and packages that up in an API that consumers can effectively use so that they don't need to know the internal implementation details of IDiagnosticService

I don't think there's any actual disagreement here between us.  It sounds like we're saying the same thing.  What i'm disagreeing with then is how we execute on this problem.  Specifically, the the execution here in this PR seems ot add much too much complexity, and far too much possibility for races to enter.  

Given that we both seem to recognize and acknowledge at least one of the root problems here, all i'm saying is that i think that root problem should have a dedicated clean, simple, reliable, and well tested solution provided.  Given how deeply coupled this will be to hte behavior of IDiagnosticService, and given how IDiagnosticService itself already has deep knowledge and behavioral understanding of WorkspaceEvents, my suggestion would be that someone with deep knowledge of both work on this.

CyrusNajmabadi commented:
> which is a bug? who ensure tagger to be refreshed?

The point is that that shouldn't be happening directly by using this.  Code that cares about this is supposed to use something like TaggerEventSources.OnWorkspaceRegistrationChanged.

heejaechang commented:
I know it uses WorkspaceChangedEventSource.

but it doesn't actually ask for new data from diagnostic service for new workspace. existing design of tagger assumes tagger itself calculate tag data rather than there is another service that calculates for it.

ones that have other service that calculate the data and tagger is just one that pass that along, it usually doesn't use existing tagger system (ex) rename tagger. 

for diagnostics tagger, we decide to use existing system, but uses diagnostic service to get info. it is unique choice for the diagnostic tagger, so in my opinion, it must be the tagger that deal with it.

CyrusNajmabadi commented:
> for diagnostics tagger, we decide to use existing system, but uses diagnostic service to get info. it is unique choice for the diagnostic tagger,

I don't see it as a "choice" :)  .  Were there other alternatives the diagnostic tagger could have taken?  If not, there wasn't really a 'choice', but instead the tagger had to be written that way because of 'necessity'.  

Again, we decided to expose diagnostics from Roslyn in a way that is different from virtually every other data source in the system.  We've seen the consequences of this (5 years+ of ongoing bugs in this area).  IMO, not exposing a system for diagnostics that behaves like all our other data source APIs is the root problem here.

CyrusNajmabadi commented:
>  existing design of tagger assumes tagger itself calculate tag data rather than there is another service that calculates for it.

That's not actually the design.  The design requirement of the tagger architecture is only that *after* receiving an event, it should be possible to ask for the data necessary to produce all the tags.  The tagger infrastructure then takes it from there.  Given the above, it can always achieve a fixed point as events come in.  It can also efficiently process, diff, the new data against the last successful stored snapshot of data it processed.  

This works well in roslyn as we've designed so many of our systems to be able to be used in such a manner.  Namely "hey, something changed! (possibly with the changed data)" and "hey, give me the data you have so i can decide what to do".  This pattern works extremely well and has provided virtually bug free subsystems (at least wrt to eventing/asynchrony/concurrency/data-management) for years.  The only outstanding bug (which is really a feature request) we have in this area is to expose a little more flexibility in this system around the timings the tagger system using when doing this processing.

heejaechang commented:
retest windows_release_vs-integration_prtest please

heejaechang commented:
@dotnet/roslyn-analysis @dotnet/roslyn-ide can someone take a look?

heejaechang commented:
@CyrusNajmabadi well. we could write it its own tagger rather than implementing on top of existing tagger system we built.

"we decided to expose diagnostics from Roslyn in a way that is different from virtually every other data source in the system."

and why that is problem? it is a problem since it is different and try to jam into complex system that is not designed for it. but if it is different and have built one for it, why is that an issue?

question on why it is different at the first place is because of the necessity.

1. we can show staled data (we already had tagger with staled data due to build errors and enc) since recalculating data when something changed is not feasible for diagnostics. it has its own driver which updates diagnostics, not from tagger 
2. we have a lot of analyzers by default (10+) and in real world (roslyn IDE projects have more than 30+ analyzers) or source of diagnostics (build, enc, host diagnostics such as crash, design time build failures and etc), we could create 1 tagger for each source or somehow work with editor to let the tagger to dynamically added and removed. but we choose to just have 1 tagger with bunch of sub taggers so that we can do that ourselves.

we had different requirement, so it works differently. not sure why that makes it wrong. like I said, we could just have made it its own tagger not using existing system. like rename tagger or syntax colorizer tagger and all those other ones that doesn't fit into existing tagger system.

heejaechang commented on src/Workspaces/Core/Portable/Workspace/Workspace_Registration.cs:
```diff
@@ -58,12 +55,24 @@ protected void UnregisterText(SourceTextContainer textContainer)
+            // workspace change event is synchronous events. and it doesn't use workspace events queue since 
+            // WorkspaceRegistration is not associated with 1 specific workspace. workspace events queue is specific to 1 workspace
+            // using multiple queues that belong to multiple workspace will cause events to be out of order
+            registration.RaiseEvents(oldWorkspace, newWorkspace);
```
i have 2 choice here. creating WorkspaceRegistration's own event queue to make it async events. or make these just synchronous events like I did here.

existing code was issue since not only one of them was async and the other is not, but because it used event queue of 1 workspace which can't guarantee event ordering of WorkspaceRegistration.WorkspaceChanged events.

heejaechang commented:
@dotnet/roslyn-infrastructure what would be way to investigate this issue? https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_release_vs-integration_prtest/8712/testReport/junit/Roslyn.VisualStudio.IntegrationTests.Workspace/WorkspacesNetCore/RenamingOpenFiles/

looks like due to how I changed WorkspaceRegistration.WorkspaceChanged event.

heejaechang commented:
@dotnet/roslyn-infrastructure ping? any help on how to run integration test locally?

CyrusNajmabadi commented:
I think someone else (@jasonmalinowski and @sharwell ) should take a look.  I have not been able to convince myself of the correctness of this approach (even after pulling down the code to look at locally).  I think this need a couple (or more) people sitting together to walk through all the data sharing,  asynchrony and BG computation to try to document what's going on, and to get confidence that this code is handling it properly.  i.e. similar to how Jason and I have handled changes to AsyncLazy in the past.

I would do this, but clearly working SxS isn't something feasible right now :)  

--

NOte: because of the enormous complexity here, i would reiterate that my preferred approach here would be to extract out all that complexity into a separate component (it can live at the tagging layer if that's what you want).  That component would then be much easier to review and understand.  The tagger could then integrate with that component separately and would then manage anything UI related.  This separation between aggregation+filtering of events and the processing and UI coordination would make things *much* easier to understand and maintain IMO.

sharwell commented:
:memo: (Note to self)

Essentially all of the pending work when this bug occurs comes from the same place:
https://github.com/dotnet/roslyn/blob/1096194309547507cb3f1b8d2fcc3074c8432bab/src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs#L213

The containing type of that method is generic. Three different instantiations each contribute:
 
* `AggregatingTagger<IErrorTag>`
* `AggregatingTagger<SuggestionTag>`
* `AggregatingTagger<ClassificationTag>`

CyrusNajmabadi commented:
> Essentially all of the pending work when this bug occurs comes from the same place:

If that's the case, then addressing this with a simpler solution is definitely possible.  We just need to take the approach where we see if there a pending foreground notification already in flight, and not bother to make a new one if so.  And instead of passing the argument along in the lambda, it just buffered into a list of work to do.  When the foreground notification actually activates, it pulls off all the work, marks that any incoming work should cause a new foreground notification, and then processes it all.

Tihs way, there is only one outstanding pending foreground request at a time.  Or, in this case, 3 (since there are three types).  

This sort of pattern is simple to do, and we've done it in many different places.  I'd be happy to contribute such a fix later today or tomorrow.

heejaechang commented on src/Features/Core/Portable/Diagnostics/DiagnosticService.cs:
```diff
@@ -55,16 +56,63 @@ public DiagnosticService([ImportMany] IEnumerable<Lazy<IAsynchronousOperationLis
             }
         }
 
+        public IDisposable Subscribe(Workspace workspace, DocumentId documentId, Action<DiagnosticsUpdatedArgs> action)
```
this

jasonmalinowski commented on src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs:
```diff
@@ -47,55 +49,54 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
-
-                _taskChain = task;
+                // make sure we belong to same workspace
+                Contract.ThrowIfFalse(document == null || _registration.Workspace == document.Project.Solution.Workspace);
```
My first reaction was "that could never happen" and this could be deleted. But then again I guess that's why to leave it... :-)

heejaechang commented:
closing in favor of https://github.com/dotnet/roslyn/pull/23448
